From 1ce7378be59367726ddc484c1ad5535fe8e11fe4 Mon Sep 17 00:00:00 2001
From: Varsha Maira <vmaira@habana.ai>
Date: Thu, 20 Jun 2024 13:23:04 +0300
Subject: [PATCH] [FSW-418] fs1: bootrom: allow configurable tpm spi freq

In FS1, TPM SPI frequency is configured based on a
2-bit setting in eFuse. This patch adds changes to
read frequency configuration from efuse and compute
the baudrate to be configured to SPI controller
accordingly.

Change-Id: Ic920914755f88bede4856ac7af6ffea5b7d8b535
Signed-off-by: Varsha Maira <vmaira@habana.ai>
---

diff --git a/arch/fs1/arc/asic.c b/arch/fs1/arc/asic.c
index 46d8e23..4a23c15 100644
--- a/arch/fs1/arc/asic.c
+++ b/arch/fs1/arc/asic.c
@@ -11,6 +11,9 @@
 #include <fs1_blocks_linux_driver.h>
 #include <pll_control_regs.h>
 #include <dev_efuse.h>
+#include <fs1_efuse.h>
+#include <common.h>
+#include <tpm.h>
 
 #define BOOTROM_UART_BAUDRATE	115200
 
@@ -61,3 +64,22 @@
 
 	return false;
 }
+
+uint32_t boot_rom_tpm_spi_baudr_get(void)
+{
+	enum boot_rom_tpm_spi_freq_config freq_config;
+	uint32_t baudr;
+
+	freq_config = boot_rom_get_tpm_spi_freq_config();
+
+	if (freq_config == FREQ_CFG_12_5MHZ)
+		baudr = (SYS_REF_CLOCK / SPI_TPM_FREQ_12_5M);
+	else
+		baudr = (SYS_REF_CLOCK / SPI_TPM_FREQ_6_25M);
+
+	/* baud rate doesn't support odd number and must not be 0 */
+	baudr = (baudr < 2) ? 2 : baudr;
+	baudr = (baudr + 1) & 0xFFFE;
+
+	return baudr;
+}
diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 18f5c7c..25b675c 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -315,3 +315,13 @@
 		PROGRESS_BOOT_ROM_SVN_TPM_PCR_EXTEND_FAIL :
 		PROGRESS_BOOT_ROM_MANIFEST_TPM_PCR_EXTEND_FAIL;
 }
+
+enum boot_rom_tpm_spi_freq_config boot_rom_get_tpm_spi_freq_config(void)
+{
+	enum boot_rom_tpm_spi_freq_config freq_config;
+
+	freq_config = ((boot_rom_global.config_fuse & EFUSE_SYSCFG_SPI_TPM_BAUDR) == 0) ?
+		FREQ_CFG_12_5MHZ : FREQ_CFG_6_25MHZ;
+
+	return freq_config;
+}
diff --git a/arch/fs1/arc/onix.c b/arch/fs1/arc/onix.c
index f696abd..ee3aaa6 100644
--- a/arch/fs1/arc/onix.c
+++ b/arch/fs1/arc/onix.c
@@ -5,7 +5,10 @@
  */
 
 #include <boot_rom.h>
+#include <fs1_efuse.h>
 #include "dev_uart.h"
+#include <common.h>
+#include <tpm.h>
 
 #define ONIX_REF_CLK		35000000ULL
 #define ONIX_UART_BAUDRATE	57600
@@ -24,3 +27,22 @@
 {
 	return true;
 }
+
+uint32_t boot_rom_tpm_spi_baudr_get(void)
+{
+	enum boot_rom_tpm_spi_freq_config freq_config;
+	uint32_t baudr;
+
+	freq_config = boot_rom_get_tpm_spi_freq_config();
+
+	if (freq_config == FREQ_CFG_12_5MHZ)
+		baudr = (ONIX_REF_CLK / SPI_TPM_FREQ_12_5M);
+	else
+		baudr = (ONIX_REF_CLK / SPI_TPM_FREQ_6_25M);
+
+	/* baud rate doesn't support odd number and must not be 0 */
+	baudr = (baudr < 2) ? 2 : baudr;
+	baudr = (baudr + 1) & 0xFFFE;
+
+	return baudr;
+}
diff --git a/arch/fs1/arc/pldm.c b/arch/fs1/arc/pldm.c
index 5fbef2f..1935054 100644
--- a/arch/fs1/arc/pldm.c
+++ b/arch/fs1/arc/pldm.c
@@ -12,7 +12,10 @@
 #include <pll_control_regs.h>
 #include <uart_regs.h>
 #include <dev_efuse.h>
+#include <common.h>
+#include <tpm.h>
 
+#define SYS_CLOCK_FREQUENCY	250000000ULL /* PCI PLL DIV3 Emulation clk */
 #define PLDM_UART_REF_CLK	250000000ULL
 #define PLDM_UART_BAUDRATE	300
 
@@ -50,3 +53,22 @@
 
 	return false;
 }
+
+uint32_t boot_rom_tpm_spi_baudr_get(void)
+{
+	enum boot_rom_tpm_spi_freq_config freq_config;
+	uint32_t baudr;
+
+	freq_config = boot_rom_get_tpm_spi_freq_config();
+
+	if (freq_config == FREQ_CFG_12_5MHZ)
+		baudr = (SYS_CLOCK_FREQUENCY / SPI_TPM_FREQ_12_5M);
+	else
+		baudr = (SYS_CLOCK_FREQUENCY / SPI_TPM_FREQ_6_25M);
+
+	/* baud rate doesn't support odd number and must not be 0 */
+	baudr = (baudr < 2) ? 2 : baudr;
+	baudr = (baudr + 1) & 0xFFFE;
+
+	return baudr;
+}
\ No newline at end of file
diff --git a/common/common.h b/common/common.h
index aa2515c..707e309 100644
--- a/common/common.h
+++ b/common/common.h
@@ -31,5 +31,6 @@
 bool boot_rom_is_manifest_valid(void);
 bool boot_rom_get_hash_for_tpm_pcr(uint8_t tpm_pcr_index, uint8_t *hash);
 enum boot_rom_progress boot_rom_get_tpm_extend_fail_prog_bit(uint8_t tpm_pcr_index);
+enum boot_rom_tpm_spi_freq_config boot_rom_get_tpm_spi_freq_config(void);
 
 #endif /* _COMMON_H_ */
diff --git a/common/tpm.h b/common/tpm.h
index 127349d..f023b41 100644
--- a/common/tpm.h
+++ b/common/tpm.h
@@ -6,6 +6,11 @@
 #ifndef _TPM_H_
 #define _TPM_H_
 
+#include <boot_rom.h>
+
+#define SPI_TPM_FREQ_12_5M	(SYS_REF_CLOCK / 4)
+#define SPI_TPM_FREQ_6_25M	(SYS_REF_CLOCK / 8)
+
 #define NUM_PCR_TO_EXTEND_IN_BOOTROM	(2)
 
 void boot_rom_tpm_ops(bool long_timeout_en);
diff --git a/common/tpm2.c b/common/tpm2.c
index 4658643..4ba207b 100644
--- a/common/tpm2.c
+++ b/common/tpm2.c
@@ -18,6 +18,7 @@
 #include <stdarg.h>
 #include <dev_spi_tpm.h>
 #include <spi.h>
+#include "common.h"
 
 /* Internal error of TPM command library */
 #define TPM_LIB_ERROR ((uint32_t)~0u)
@@ -151,12 +152,6 @@
 	reg = regmap_readl(out_en);
 	reg |= mask;
 	regmap_writel(reg, out_en);
-}
-
-#define DW_TPM_SPI_BAUD_RATE		4
-uint32_t boot_rom_tpm_spi_baudr_get(void) /*TODO : get baudr from fuse setting*/
-{
-	return DW_TPM_SPI_BAUD_RATE;
 }
 
 static void boot_rom_tpm_spi_local_ctx_init(void)
